cdap-proto/src/main/java/io/cdap/cdap/proto/id/ApplicationDraftId.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.cdap.proto.id;
+
+import io.cdap.cdap.proto.element.EntityType;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Objects;
+
+/**
+ * Uniquely identifies a draft.
+ */
+public class ApplicationDraftId extends NamespacedEntityId implements ParentedId<NamespaceId> {
+  private final String id;
+  private transient Integer hashCode;
+
+  public ApplicationDraftId(String namespace, String id) {
+    super(namespace, EntityType.APPLICATION_DRAFT);
+    if (id == null) {
+      throw new NullPointerException("ID cannot be null.");
+    }
+    ensureValidId("application draft", id);
+    this.id = id;
+  }
+
+  public String getId() {
+    return id;
+  }
+
+  @Override
+  public String getEntityName() {
+    return getId();
+  }
+
+  @Override
+  public NamespaceId getParent() {
+    return new NamespaceId(namespace);
```

Comment:
Conceptually draft should be a child resource of application, right?

Reply:
When draft is created, application does not exist yet. It's a prequel to a possible application (if it's created out of draft).

Reply:
I see. 

I considered studio to be an application and draft is a resource within that application (thus below application resource type in the resource hierarchy) 

But you are also right that one can treat draft as a resource type under namespace. It is just another form of application before deployment.

Reply:
It seems like draft is a concept of applications. I dont think any cdap application can have draft. I feel like this should be enforced at app level?
@albertshau any thoughts?

Reply:
From a use case perspective, agree that the parent of a draft should be a namespace and not the pipeline studio app. Are we introducing a draft entity because we want somebody to be able to create and edit drafts, but not be able to deploy that draft into an application? Wondering why the application entity can't be used. 

I'm concerned that application specific concepts like a draft need to be added to the CDAP platform level in order for applications to enforce auth on them. Are there plans to have custom entities that applications can define themselves? For example, connections and wrangler workspaces are other application defined entities that we will need to add auth support for. If there are plans, wondering if drafts make more sense there instead of as an entity type at the platform level.